File: packages/sap-testing-automation/web-tests/journal.ts
/**
 * The run journal — what actually happened, recorded by the run itself.
 *
 * Everything else in this workspace was already model-free: the specs drive the
 * screens, `screens/*.json` holds the handles, `test-data/*.dataset.json` holds
 * the values. One thing was not. The record of a run — `results/TC-*.md`, the
 * file the dashboard reads, the freeze gate counts and the case's Run history
 * cites — was written afterwards, by hand, from a transcript. That is the last
 * place a human (or a model) stands between the system under test and the
 * report about it, and it is the worst place for one to stand: it is where
 * CLAUDE.md rules 5 and 6 are enforced by good intentions rather than by
 * mechanism.
 *
 * So the run emits its own record. A spec calls `step`, `check`, `document`
 * and `deviation` as it goes; each becomes one NDJSON line under
 * `results/web/<SYSTEM_ID>/journal/`. `reporters/result-file.ts` renders those
 * lines into the run file at the end. Nothing is transcribed, so nothing can be
 * embellished — an observed value in the report is the string that came off the
 * screen, and a field the run never read is absent here and prints as
 * `NOT OBSERVED` there, which is rule 6 as a data flow rather than as a promise.
 *
 * Three properties this file protects:
 *
 *  - **It never breaks a run.** Every entry point is wrapped: a journal that
 *    cannot write must not fail a test that is otherwise fine, because the
 *    expensive failure here is a real SAP write that gets rolled back over a
 *    reporting bug. A journal write that fails costs a thinner report.
 *  - **It never invents.** There is no defaulting, no inference and no
 *    "probably". `check` records the value it was handed; if a spec has nothing
 *    to hand it, the row says so.
 *  - **It is additive.** A spec that calls none of this still gets a run file —
 *    the reporter falls back to Playwright's own view (test titles, statuses,
 *    `expect` steps). Instrumenting a spec makes its report better; not
 *    instrumenting one does not make it silent.
 */
import { appendFileSync, mkdirSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';
import { test } from '@playwright-sap/test';
import { sapSystem } from './sap-system';

const here = dirname(fileURLToPath(import.meta.url));
const repoRoot = resolve(here, '..');

/**
 * One id per invocation, shared by every worker in it.
 *
 * `scripts/run-case.ps1` sets `SAP_RUN_ID` so the runner knows the filename
 * before Playwright starts. A bare `npx playwright test` gets a generated one,
 * so an ad-hoc run still produces a run file rather than silently skipping the
 * record.
 */
export const RUN_ID =
  process.env.SAP_RUN_ID?.trim() ||
  `${new Date().toISOString().replace(/[-:T]/g, '').slice(0, 12)}-${process.pid}`;

/** Who is answerable for this run. Recorded, never guessed at render time. */
export const RUN_BY = process.env.SAP_RUN_BY?.trim() || '';

/** Optional suffix for the run filename, e.g. `create-only`. */
export const RUN_TAG = process.env.SAP_RUN_TAG?.trim() || '';

export type Outcome = 'ok' | 'skipped' | 'error' | 'refused' | 'blocked';

/**
 * `recorded` is not a weaker `pass` — it is the absence of a verdict.
 *
 * `check` writes down a value; it does not assert anything about it. Defaulting
 * that to `pass` would put a verdict in the report that nothing in the run ever
 * checked, which is rule 6 ("never invent a result") broken by a default
 * argument. Use `checked` to get a pass or a fail, or pass one explicitly.
 */
export type CheckResult = 'pass' | 'fail' | 'not-observed' | 'recorded';

export type JournalEntry =
  | { kind: 'meta'; testId: string; at: string; case?: string; key?: string; value?: string }
  | { kind: 'step'; testId: string; at: string; description: string; outcome: Outcome; detail?: string }
  | {
      kind: 'check';
      testId: string;
      at: string;
      description: string;
      expected: string;
      observed: string | null;
      result: CheckResult;
    }
  | {
      kind: 'document';
      testId: string;
      at: string;
      docType: string;
      number: string | null;
      companyCode?: string;
      lifecycle: string[];
      leftInPlace: boolean;
      note?: string;
    }
  | { kind: 'evidence'; testId: string; at: string; file: string; shows: string }
  | { kind: 'deviation'; testId: string; at: string; text: string }
  | {
      kind: 'system';
      testId: string;
      at: string;
      where: string;
      system: string;
      client: string;
      user?: string;
      /**
       * Whether the DS4/100 check actually held. Absent on journals written
       * before this field existed, which the reporter treats as "not stated"
       * rather than inventing a verdict either way.
       */
      confirmed?: boolean;
    }
  | { kind: 'verdict'; testId: string; at: string; verdict: string; why: string };

function journalDir() {
  return resolve(repoRoot, 'results', 'web', sapSystem.id, 'journal');
}

export function journalPath(runId = RUN_ID) {
  return resolve(journalDir(), `${runId}.ndjson`);
}

/**
 * Which test a line belongs to.
 *
 * `test.info()` is the runner's own view of what is executing, so entries group
 * per test without any spec passing an id around — and it stays correct inside
 * `treasury.ts`, which is called by nine different specs and knows nothing
 * about any of them. Called outside a test (module scope, a global hook) it
 * throws; `-` groups those lines under the run itself rather than dropping
 * them.
 */
function currentTestId(): string {
  try {
    return test.info().testId;
  } catch {
    return '-';
  }
}

/**
 * Append one line. Swallows every error on purpose — see the header: a report
 * that cannot be written must not fail a test that is otherwise fine.
 */
function emit(entry: Omit<JournalEntry, 'testId' | 'at'> & Record<string, unknown>) {
  try {
    mkdirSync(journalDir(), { recursive: true });
    const line = JSON.stringify({
      ...entry,
      testId: currentTestId(),
      at: new Date().toISOString(),
    });
    appendFileSync(journalPath(), line + '\n', 'utf8');
  } catch {
    /* the run matters more than the record of it */
  }
}

/** Which case this test is a run of. Without it the reporter has no filename. */
export function forCase(caseId: string) {
  emit({ kind: 'meta', case: caseId });
}

/** A free-form fact for the run file's header, e.g. dataset row or stage. */
export function meta(key: string, value: string) {
  emit({ kind: 'meta', key, value: String(value) });
}

/** One row of the run file's "Steps executed" table. */
export function step(description: string, outcome: Outcome = 'ok', detail?: string) {
  emit({ kind: 'step', description, outcome, detail });
}

/**
 * One row of the "Assertions" table, with the value that was actually read.
 *
 * `observed` is whatever came off the screen — pass it through unchanged, even
 * when it is ugly (`100,000.00`, `10.0000000`). Reformatting it here would make
 * the report say something the system did not.
 *
 * Returns `observed` so it can wrap an existing expression, and records
 * `not-observed` for a null/undefined read rather than printing an empty cell.
 *
 * With no explicit `result` the row is `recorded`, not `pass`: this function
 * asserts nothing, and claiming a verdict it did not reach would be inventing
 * one. `checked` is the version that asserts.
 */
export function check(
  description: string,
  expected: unknown,
  observed: unknown,
  result?: CheckResult,
): string | null {
  const obs =
    observed === null || observed === undefined || observed === ''
      ? null
      : String(observed);
  emit({
    kind: 'check',
    description,
    expected: String(expected),
    observed: obs,
    result: result ?? (obs === null ? 'not-observed' : 'recorded'),
  });
  return obs;
}

/**
 * Record a check and assert it in one call, so the report and the verdict can
 * never disagree.
 *
 * The failure path matters more than the happy one: `assertFn` throwing records
 * the row as `fail` *before* the error propagates, so a run that dies on its
 * third assertion still reports the first two and the one that killed it. A
 * spec that asserts without recording leaves the reporter to infer the row from
 * Playwright's `expect` step, which knows the description but not the value.
 */
export function checked<T>(
  description: string,
  expected: unknown,
  observed: T,
  assertFn: (observed: T) => void,
): T {
  try {
    assertFn(observed);
  } catch (e) {
    check(description, expected, observed, 'fail');
    throw e;
  }
  check(description, expected, observed, 'pass');
  return observed;
}

/**
 * A business object this run wrote. One row of "Documents created".
 *
 * `number` is null when the write was attempted and no number came back — that
 * is a real, reportable state ("attempted, wrote nothing") and is not the same
 * as no row at all. `scripts/build-dashboard.ps1` distinguishes them.
 */
export function document(args: {
  docType: string;
  number: string | null;
  companyCode?: string;
  lifecycle?: string[];
  leftInPlace?: boolean;
  note?: string;
}) {
  emit({
    kind: 'document',
    docType: args.docType,
    number: args.number || null,
    companyCode: args.companyCode,
    lifecycle: args.lifecycle ?? [],
    leftInPlace: args.leftInPlace ?? true,
    note: args.note,
  });
}

/** Extend a document's lifecycle after a later step, e.g. settled -> posted. */
export function documentReached(number: string, stage: string) {
  emit({ kind: 'document', docType: '', number, lifecycle: [stage], leftInPlace: true });
}

/** One row of "Evidence". `file` is repo-relative. */
export function evidence(file: string, shows: string) {
  emit({ kind: 'evidence', file, shows });
}

/**
 * Something that differed from the case.
 *
 * This is the field the freeze gate reads: `scripts/check-suite.ps1` refuses to
 * count a PASS run that recorded a deviation toward the two clean runs a case
 * needs to freeze. Recording one therefore costs something, which is exactly
 * why it has to be the run that records it and not a narrator deciding later
 * whether the popup "really counted".
 */
export function deviation(text: string) {
  emit({ kind: 'deviation', text });
}

/**
 * The system confirmation behind CLAUDE.md rule 1, as a recorded fact.
 *
 * `confirmed` is the verdict of the DS4/100 check, not the fact that a check was
 * attempted. Callers record the failing case too — what the screen said is
 * evidence — so the reporter must read this flag rather than treating the
 * presence of an entry as a pass.
 */
export function systemConfirmed(
  where: string,
  info: { system: string; client: string; user?: string; confirmed?: boolean },
) {
  emit({
    kind: 'system',
    where,
    system: info.system,
    client: info.client,
    user: info.user,
    confirmed: info.confirmed,
  });
}

/**
 * Override the verdict the reporter would otherwise derive.
 *
 * Only for the verdicts Playwright cannot see. A precondition that was not met
 * is `BLOCKED`, not a failure of the product, and a batch that wrote six of ten
 * rows is `PARTIAL` even though every test in it passed — neither is visible
 * from a green test result, and both change what the dashboard says.
 */
export function verdict(v: 'PASS' | 'FAIL' | 'BLOCKED' | 'PARTIAL', why: string) {
  emit({ kind: 'verdict', verdict: v, why });
}

export const journal = {
  RUN_ID,
  RUN_BY,
  RUN_TAG,
  forCase,
  meta,
  step,
  check,
  checked,
  document,
  documentReached,
  evidence,
  deviation,
  systemConfirmed,
  verdict,
};

File: packages/sap-testing-automation/gui_tests/journal.py
"""
The GUI lane's run journal — what actually happened, recorded by the run itself.

This is `web-tests/journal.ts` for the other lane, deliberately emitting the
**same NDJSON schema**. That is the whole design decision: one schema means
`scripts/check-run.ps1` reads a GUI-lane run the same way it reads a web-lane
one, `scripts/build-dashboard.ps1` counts it the same way, and the run file
renderer is one implementation rather than two that drift.

The rules it inherits, and why they matter more here than in the web lane:

  - **It never breaks a run.** Every entry point swallows its own errors. A
    journal that cannot write must not fail a run that is otherwise fine —
    the expensive failure is a real SAP write abandoned over a reporting bug.
  - **It never invents.** `check()` records the value it was handed. Nothing
    defaults, nothing infers. A field the run never read is absent here and
    prints `NOT OBSERVED` in the run file, which is CLAUDE.md rule 6 as a data
    flow rather than as a promise.
  - **It is append-and-flush, one line per action.** This is what makes an
    interrupted run recoverable: a killed process loses at most the line it was
    mid-write on, never what came before. TC-014's COM disconnect is exactly
    the scenario — see `docs/unattended-runs.md` § Recovering from an
    interrupted run.

Unlike the web lane there is no test runner underneath this, so there is no
`test.info()` to group entries by. Everything in one GUI-lane run belongs to
one case, so `testId` is a constant.
"""
from __future__ import annotations

import json
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Iterable, Sequence

REPO_ROOT = Path(__file__).resolve().parent.parent

#: The GUI lane's journals live beside the web lane's, one directory over.
#: `scripts/check-run.ps1` scans both roots.
JOURNAL_ROOT = REPO_ROOT / "results" / "gui"

#: Every entry in a GUI-lane run belongs to the one case being run. The field
#: exists only so the schema matches the web lane's, where a run can interleave
#: several tests.
_TEST_ID = "gui"

OUTCOMES = ("ok", "skipped", "error", "refused", "blocked")
CHECK_RESULTS = ("pass", "fail", "not-observed", "recorded")


def _now() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


class Journal:
    """One run's journal. Append-only, flushed per line."""

    def __init__(self, run_id: str, system_id: str, case_id: str | None = None,
                 run_by: str = "", run_tag: str = ""):
        self.run_id = run_id
        self.system_id = system_id
        self.case_id = case_id
        self.run_by = run_by
        self.run_tag = run_tag
        self.path = JOURNAL_ROOT / system_id / "journal" / f"{run_id}.ndjson"
        if case_id:
            self.for_case(case_id)

    # ---------------------------------------------------------------- emit

    def _emit(self, **entry: Any) -> None:
        """Append one line. Swallows every error on purpose — see the header."""
        try:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            entry["testId"] = _TEST_ID
            entry["at"] = _now()
            line = json.dumps(entry, ensure_ascii=False, default=str)
            # Open/append/close per line so a hard kill cannot lose a buffer.
            with self.path.open("a", encoding="utf-8") as fh:
                fh.write(line + "\n")
                fh.flush()
                os.fsync(fh.fileno())
        except Exception:
            pass  # the run matters more than the record of it

    # -------------------------------------------------------------- writers

    def for_case(self, case_id: str) -> None:
        """Which case this run is of. Without it the renderer has no filename."""
        self.case_id = case_id
        self._emit(kind="meta", case=case_id)

    def meta(self, key: str, value: Any) -> None:
        """A free-form fact for the run file's header, e.g. dataset row."""
        self._emit(kind="meta", key=key, value=str(value))

    def step(self, description: str, outcome: str = "ok", detail: str | None = None) -> None:
        """One row of the run file's "Steps executed" table."""
        self._emit(kind="step", description=description, outcome=outcome, detail=detail)

    def check(self, description: str, expected: Any, observed: Any,
              result: str | None = None) -> str | None:
        """
        One row of the "Assertions" table, with the value actually read.

        `observed` is passed through unchanged, even when it is ugly
        (`100,000.00`, `10.0000000`) — reformatting would make the report say
        something the system did not.

        With no explicit `result` the row is `recorded`, not `pass`: this
        records, it does not assert. Use `checked()` for the version that
        asserts, so report and verdict cannot disagree.
        """
        obs = None if observed is None or observed == "" else str(observed)
        self._emit(
            kind="check",
            description=description,
            expected=str(expected),
            observed=obs,
            result=result or ("not-observed" if obs is None else "recorded"),
        )
        return obs

    def checked(self, description: str, expected: Any, observed: Any,
                predicate=None) -> Any:
        """
        Record and assert in one call.

        The failure path is the point: a failing predicate records the row as
        `fail` *before* raising, so a run that dies on its third assertion still
        reports the first two and the one that killed it.
        """
        if predicate is None:
            def predicate(o):
                return str(o) == str(expected)
        try:
            ok = predicate(observed)
        except Exception:
            self.check(description, expected, observed, "fail")
            raise
        self.check(description, expected, observed, "pass" if ok else "fail")
        if not ok:
            raise AssertionError(
                f"{description}: expected {expected!r}, observed {observed!r}"
            )
        return observed

    def document(self, doc_type: str, number: str | None, company_code: str = "",
                 lifecycle: Sequence[str] = (), left_in_place: bool = True,
                 note: str = "") -> None:
        """
        A business object this run wrote. One row of "Documents created".

        `number=None` is a real, reportable state — "attempted, wrote nothing" —
        and is not the same as no row at all.
        """
        self._emit(
            kind="document",
            docType=doc_type,
            number=number or None,
            companyCode=company_code,
            lifecycle=list(lifecycle),
            leftInPlace=left_in_place,
            note=note,
        )

    def document_reached(self, number: str, stage: str) -> None:
        """Extend a document's lifecycle after a later step, e.g. settled -> posted."""
        self._emit(kind="document", docType="", number=number,
                   lifecycle=[stage], leftInPlace=True)

    def evidence(self, file: str, shows: str) -> None:
        """One row of "Evidence". `file` is repo-relative."""
        self._emit(kind="evidence", file=file, shows=shows)

    def deviation(self, text: str) -> None:
        """
        Something that differed from the case.

        This is the field the freeze gate reads: `scripts/check-suite.ps1`
        refuses to count a PASS run that recorded a deviation toward the two
        clean runs a case needs to freeze. Recording one therefore costs
        something — which is exactly why the run records it, rather than a
        narrator deciding afterwards whether the popup "really counted".
        """
        self._emit(kind="deviation", text=text)

    def system_confirmed(self, where: str, system: str, client: str,
                         user: str = "", confirmed: bool | None = None) -> None:
        """
        CLAUDE.md rule 1 as a recorded fact.

        `confirmed` is the verdict of the DS4/100 check, not the fact that a
        check happened. The failing case is recorded too — what the screen said
        is evidence — so readers must read the flag, not the row's existence.
        """
        self._emit(kind="system", where=where, system=system, client=client,
                   user=user, confirmed=confirmed)

    def verdict(self, verdict: str, why: str) -> None:
        """
        Override the verdict the renderer would otherwise derive.

        Only for what the runner cannot see: an unmet precondition is `BLOCKED`,
        not a product failure; a batch that wrote six of ten rows is `PARTIAL`
        even with no exception raised.
        """
        self._emit(kind="verdict", verdict=verdict, why=why)


def read_journal(path: Path) -> list[dict]:
    """
    Parse a journal, tolerating a torn last line.

    A kill mid-write loses at most one line; losing the rest of the file with it
    would defeat the reason the journal exists. Same tolerance as
    `web-tests/reporters/result-file.ts` and `scripts/check-run.ps1`.
    """
    if not path.exists():
        return []
    out: list[dict] = []
    for raw in path.read_text(encoding="utf-8").splitlines():
        line = raw.strip()
        if not line:
            continue
        try:
            out.append(json.loads(line))
        except Exception:
            pass
    return out
